File: Model/matrix_form/lstm_inference.py
import numpy as np
import time, tracemalloc


def sigmoid(x: np.ndarray) -> np.ndarray:
    return 1 / (1 + np.exp(-x))


def lstm_cell(
    x: np.ndarray,
    h: np.ndarray,
    c: np.ndarray,
    W_ih: np.ndarray,
    W_hh: np.ndarray,
    b_h: np.ndarray,
) -> tuple[np.ndarray, np.ndarray]:
    """
    Perform a single step of the LSTM cell.

    Args:
        x (numpy.ndarray): Input tensor of shape (batch_size, input_size).
        hidden (tuple): Tuple containing the previous hidden state (h) and cell state (c).
        W_ih (numpy.ndarray): Input-to-hidden weight matrix of shape (hidden_size, input_size).
        W_hh (numpy.ndarray): Hidden-to-hidden weight matrix of shape (hidden_size, hidden_size).
        b_h (numpy.ndarray): Hidden-to-hidden bias vector of shape (hidden_size,).

    Returns:
        tuple: Tuple containing the updated hidden state (h) and cell state (c).
    """
    gates = np.dot(W_ih, x) + np.dot(W_hh, h) + b_h
    i, f, g, o = np.split(gates, 4)
    i = sigmoid(i)
    f = sigmoid(f)
    g = np.tanh(g)
    o = sigmoid(o)
    c = f * c + i * g
    h = o * np.tanh(c)
    return h, c


def lstm_matrix_inference(
    W_xh: list[np.ndarray], W_hh: np.ndarray, b_h: np.ndarray, batch_size: int
):
    """
    Create a function that performs forward pass through LSTM layers.

    Args:
        W_xh (list[np.ndarray]): List of weight matrices for input-to-hidden connections for each LSTM layer.
        W_hh (np.ndarray): Weight matrix for hidden-to-hidden connections.
        b_h (np.ndarray): Bias vector for hidden-to-hidden connections.
        batch_size (int): Batch size.

    Returns:
        function: A function that performs forward pass through LSTM layers.
    """
    num_layers = len(W_xh)
    hidden_size = W_hh.shape[-1]

    def forward(x: np.ndarray) -> np.ndarray:
        """
        Perform forward pass through the LSTM layers.

        Args:
            x (np.ndarray): Input sequence of shape (seq_len, input_size).

        Returns:
            np.ndarray: Output hidden state of the last LSTM layer.
        """
        hidden = np.zeros((num_layers, hidden_size))
        cell = np.zeros((num_layers, hidden_size))
        seq_len = x.shape[0]
        for t in range(seq_len):
            hidden[0], cell[0] = lstm_cell(x[t], hidden[0], cell[0], W_xh[0], W_hh[0], b_h[0])
            hidden[1], cell[1] = lstm_cell(hidden[0], hidden[1], cell[1], W_xh[1], W_hh[1], b_h[1])
        return hidden[1]

    return forward


def fc_matrix_inference(W: np.ndarray, b: np.ndarray):
    """
    Create a function for performing forward inference using the fully connected matrix operation.

    Args:
        W (np.ndarray): The weight matrix of shape (n_outputs, n_features).
        b (np.ndarray): The bias vector of shape (n_outputs,).

    Returns:
        function: A function that performs the forward inference operation.
    """
    def forward(input_):
        """
        Perform forward inference using the fully connected matrix operation.

        Args:
            input_ (np.ndarray): The input data of shape (batch_size, n_features).

        Returns:
            np.ndarray: The output of the forward inference operation of shape (batch_size, n_outputs).
        """
        return np.dot(W, input_) + b

    return forward


def npiter(arr: np.ndarray):
    """
    Iterate over a numpy array.

    Args:
        arr (np.ndarray): The input numpy array.
    """
    for i in arr:
        yield i

def read_model_weights():
    weights = np.load("Model/matrix_form/LSTMRegressor.npy")
    W_hh = weights["W_hh"].squeeze()
    b_h = weights["b_h"].squeeze()
    W = weights["W"].squeeze()
    b = weights["b"].squeeze()
    W_xh = []
    for i in range(W_hh.shape[0]):
        W_xh.append(weights[f"W_xh_{i}"].squeeze())
        
    print([i.shape for i in W_xh], W_hh.shape, b_h.shape, W.shape, b.shape)

    return W_xh, W_hh, b_h, W, b

def lstm_inference(length, iterator):

    W_xh, W_hh, b_h, W, b = read_model_weights()

    batch_size = 1

    lstm_forward = lstm_matrix_inference(W_xh, W_hh, b_h, batch_size)
    fc_forward = fc_matrix_inference(W, b)

    output = np.zeros((length))
    start_time = time.perf_counter()
    for i, x in enumerate(iterator):
        output[i] = fc_forward(lstm_forward(x))
    end_time = time.perf_counter()

    print(f"Time taken for inference: {end_time - start_time:.6f} seconds")
    print((end_time - start_time) / length)
    return output

if __name__ == "__main__":
    test_array = np.load("Model/matrix_form/test_array.npy")
    target_array = np.loadtxt("Model/matrix_form/target_array.txt")
    length = test_array.shape[0]
    iterator = npiter(test_array)
    del test_array
    
    tracemalloc.start()
    output = lstm_inference(length, iterator)

    max_memory_usage = tracemalloc.get_traced_memory()[1]
    print(f"Maximum memory usage: {max_memory_usage / (1024):.2f} KB")

    tracemalloc.stop()

    print(output.shape, target_array.shape)
    from sklearn.metrics import r2_score # type: ignore
    r2 = r2_score(target_array, output)
    print(f"R2 Score: {r2}")
    for i in range(10):
        print(output[i], target_array[i])

File: Model/matrix_form/lstm_inference.cpp
#include <iostream>
#include <fstream>
#include <Eigen/Core>
#include <vector>
#include <tuple>
#include <chrono>

using namespace Eigen;

/*
(3804, 40, 2)
[(64, 2), (64, 16)] (2, 64, 16) (2, 64) (1, 16) (2,)
*/

std::vector<std::string> splitString(const std::string &str, char delimiter)
{
    std::vector<std::string> tokens;
    std::string token;
    std::stringstream ss(str);

    while (std::getline(ss, token, delimiter))
    {
        tokens.push_back(token);
    }

    return tokens;
}

Matrix<double, 64, 2> W_1h;
Matrix<double, 64, 16> W_2h;
std::vector<MatrixXd> W_hh(2, MatrixXd(64, 16));
Matrix<double, 2, 64> b_h;
Matrix<double, 1, 16> W;
Matrix<double, 2, 2> b;
std::vector<MatrixXd> test_array(3804, MatrixXd(40, 2));

void load_test_array()
{
    std::ifstream file("test_array1.txt");
    std::string line;
    for (int i = 0; i < 3804; i++)
    {
        for (int j = 0; j < 40; j++)
        {
            std::getline(file, line);
            std::vector<std::string> tokens = splitString(line, ',');
            for (int k = 0; k < 2; k++)
            {
                test_array[i](j, k) = std::stod(tokens[k]);
            }
        }
    }
    file.close();
}

void load_model_weights()
{
    std::ifstream file("LSTMRegressorWeights_strap_42.txt");
    std::string line;
    for (int i = 0; i < 2; i++)
    {
        for (int j = 0; j < 64; j++)
        {
            std::getline(file, line);
            std::vector<std::string> tokens = splitString(line, ',');
            for (int k = 0; k < 16; k++)
            {
                W_hh[i](j, k) = std::stod(tokens[k]);
            }
        }
    }

    for (int i = 0; i < 2; i++)
    {
        std::getline(file, line);
        std::vector<std::string> tokens = splitString(line, ',');
        for (int j = 0; j < 64; j++)
        {
            b_h(i) = std::stod(tokens[j]);
        }
    }

    std::getline(file, line);
    std::vector<std::string> tokens = splitString(line, ',');
    for (int i = 0; i < 16; i++)
    {
        W(0, i) = std::stod(tokens[i]);
    }

    std::getline(file, line);
    b(0, 0) = std::stod(line);

    for (int i = 0; i < 64; i++)
    {
        std::getline(file, line);
        tokens = splitString(line, ',');
        for (int j = 0; j < 2; j++)
        {
            W_1h(i, j) = std::stod(tokens[j]);
        }
    }

    for (int i = 0; i < 64; i++)
    {
        std::getline(file, line);
        tokens = splitString(line, ',');
        for (int j = 0; j < 16; j++)
        {
            W_2h(i, j) = std::stod(tokens[j]);
        }
    }
}

MatrixXd sigmoid(MatrixXd x)
{
    return 1 / (1 + (-x).array().exp());
}

MatrixXd tanh(MatrixXd x)
{
    return x.array().tanh();
}

std::tuple<MatrixXd, MatrixXd> calc_lstm_cell(MatrixXd x, MatrixXd h, MatrixXd c, MatrixXd W_ih, MatrixXd W_hh, MatrixXd b_h)
{
    MatrixXd gates = x * W_ih.transpose() + h * W_hh.transpose() + b_h;
    MatrixXd i = sigmoid(gates.block(0, 0, 1, 16));
    MatrixXd f = sigmoid(gates.block(0, 16, 1, 16));
    MatrixXd g = tanh(gates.block(0, 32, 1, 16));
    MatrixXd o = sigmoid(gates.block(0, 48, 1, 16));
    c = f.array() * c.array() + i.array() * g.array();
    h = o.array() * c.array().tanh();
    return std::make_tuple(h, c);
}

MatrixXd lstm_forward(Matrix<double, 40, 2> x)
{
    const int hidden_size = 16;
    MatrixXd hidden = MatrixXd::Zero(2, hidden_size);
    MatrixXd cell = MatrixXd::Zero(2, hidden_size);
    int seq_len = x.rows();
    for (int t = 0; t < seq_len; t++)
    {
        auto [new_h0, new_c0] = calc_lstm_cell(x.row(t), hidden.row(0), cell.row(0), W_1h, W_hh[0], b_h.row(0));
        cell.row(0) = new_c0;
        hidden.row(0) = new_h0;

        auto [new_h1, new_c1] = calc_lstm_cell(hidden.row(0), hidden.row(1), cell.row(1), W_2h, W_hh[1], b_h.row(1));
        cell.row(1) = new_c1;
        hidden.row(1) = new_h1;
    }
    return hidden.row(1);
}

MatrixXd fc_forward(Matrix<double, 1, 16> x)
{
    return x * W.transpose() + b;
}

int main()
{
    load_test_array();
    load_model_weights();


    auto start = std::chrono::high_resolution_clock::now();

    Vector<double, 3804> output;
    int counter = 0;
    for (auto& i : test_array)
    {
        output[counter] = fc_forward(lstm_forward(i))(0);
        counter++;
    }

    auto end = std::chrono::high_resolution_clock::now();

    // std::cout << output << std::endl
    //           << std::endl;

    std::cout << "Time taken: " << std::chrono::duration_cast<std::chrono::milliseconds>(end - start).count() / 1000.0 << "s" << std::endl;

    return 0;
}